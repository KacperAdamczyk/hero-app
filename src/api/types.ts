import { gql } from 'apollo-boost';

export const TYPES = gql`
  query Types {
    types {
      id
      name
    }
  }
`;
