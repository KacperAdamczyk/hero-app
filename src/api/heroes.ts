import { gql } from 'apollo-boost';

export interface HeroesParams {
  first?: number;
  skip: number;
}

export const HEROES = gql`
  query Heroes($first: Int = 8, $skip: Int!) {
    heroes(first: $first, skip: $skip) {
      id
      full_name
      avatar_url
      description
      type {
        name
      }
    }
  }
`;
