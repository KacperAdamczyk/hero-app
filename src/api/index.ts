export * from './apollo';
export * from './heroes';
