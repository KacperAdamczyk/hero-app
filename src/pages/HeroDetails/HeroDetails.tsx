/** @jsx jsx */
import { FC, useCallback } from 'react';
import { jsx, css } from '@emotion/core';
import { useNavigate, useParams } from 'react-router-dom';
import { useQuery, useMutation } from '@apollo/react-hooks';
import styled from '@emotion/styled';
import DeleteIcon from '@material-ui/icons/Delete';

import { Modal, ContentLoader, Avatar, Button, Error } from 'components';
import {
  HERO,
  Hero,
  HeroVariables,
  DELETE_HERO,
  DeleteHeroVariables,
  DeleteHero,
} from 'api';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  & > * {
    margin-bottom: 20px;
  }
`;

export const HeroDetails: FC = () => {
  const navigate = useNavigate();
  const { id } = useParams();
  const onClose = useCallback(() => navigate('/heroes'), [navigate]);

  const { loading: gettingHero, error: gettingHeroError, data } = useQuery<
    Hero,
    HeroVariables
  >(HERO, {
    variables: { id },
  });
  const [
    deleteHero,
    { loading: deletingHero, error: deletingHeroError },
  ] = useMutation<DeleteHero, DeleteHeroVariables>(DELETE_HERO);

  const onClick = useCallback(
    () =>
      deleteHero({ variables: { id } }).then(
        navigate('/heroes', { state: { refetch: true } }),
        e => console.log(e),
      ),
    [deleteHero, navigate, id],
  );

  const hero = data?.hero;

  return (
    <Modal isOpen onClose={onClose}>
      <ContentLoader loading={gettingHero} error={gettingHeroError?.message}>
        {!!hero && (
          <Container>
            <Avatar url={hero.avatar_url} size={95} />
            <div
              css={css`
                display: flex;
                flex-direction: column;
                align-items: center;
              `}
            >
              <b
                css={css`
                  font-size: 1.125em;
                  margin-bottom: 3px;
                `}
              >
                {hero.full_name}
              </b>
              <div>{hero.type.name}</div>
            </div>
            <div>{hero.description}</div>
            <Button
              color="red"
              Icon={DeleteIcon}
              loading={deletingHero}
              onClick={onClick}
            >
              Delete the hero
            </Button>
            <Error message={deletingHeroError?.message} />
          </Container>
        )}
      </ContentLoader>
    </Modal>
  );
};
