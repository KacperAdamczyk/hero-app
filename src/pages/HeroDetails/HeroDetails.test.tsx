import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/react-testing';
import { MemoryRouter } from 'react-router-dom';
import waitForExpect from 'wait-for-expect';

import { HeroDetails } from './HeroDetails';
import { HERO } from 'api';

jest.mock('react-modal');

let queryFetched = false;

const query: MockedResponse[] = [
  {
    request: {
      query: HERO,
    },
    result: () => {
      queryFetched = true;

      return {
        data: {
          hero: {
            __typename: 'hero',
            avatar_url: 'avatar',
            full_name: 'name',
            type: {
              name: 'type',
            },
            description: 'description',
          },
        },
      };
    },
  },
];

describe('HeroDetails', () => {
  beforeEach(() => {
    queryFetched = false;
  });

  test('should query a hero', async () => {
    const { container } = render(
      <MockedProvider mocks={query}>
        <MemoryRouter>
          <HeroDetails />
        </MemoryRouter>
      </MockedProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
    await waitForExpect(() => {
      expect(queryFetched).toBeTruthy();
    });
  });
});
