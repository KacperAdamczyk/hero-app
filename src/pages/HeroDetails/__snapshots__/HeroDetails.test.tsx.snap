// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroDetails should query a hero 1`] = `null`;

exports[`HeroDetails should query data 1`] = `null`;

exports[`HeroDetails should query data 2`] = `null`;
